### PR TITLE
Dask integration

### PR DESCRIPTION
**Pull request by jim22k**
This will give you a DaskResolver object wrapping the default `mg.resolver`.
```python
from metagraph.dask import resolver
```
Use it as you would the normal resolver, but returned objects are dask objects, requiring `.compute()`.

To wrap you own custom resolver, use:
```python
from metagraph.dask import DaskResolver
lazy = DaskResolver(my_custom_resolver)
```

Calling algorithms and calling `.translate` will return `Placeholder` objects for each ConcreteType. This allows the system to know the input type when choosing translators and concrete algorithms.

Translations as part of an algorithm call and multi-step translations are converted into separate dask tasks.

Tests now have a `--dask` calling option which will run all tests using a `DaskResolver`.

### CODE REVIEW

File: metagraph/core/dask/resolver.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        if ct is None:
+            ct = self._resolver.class_to_concrete.get(klass)
+            if ct is None:
+                raise TypeError(
+                    f"{klass.__name__} is not a defined `value_type`. Must provide `concrete_type`."
+                )
+        ph = self._get_placeholder(ct)
+        return DelayedWrapper(klass, ph)
+
+    def _get_placeholder(self, concrete_type):
```
Comment (seibert) [inline metagraph/core/dask/resolver.py:77]: Can you put some doc strings explaining what the internal methods on this class are doing?  I think I get it from reading the body, but this code is a little hard to follow at first.

File: metagraph/core/dask/resolver.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        # Walk through arguments and apply required translations
+        bound = sig.bind(*args, **kwargs)
+        bound.apply_defaults()
+        for name, trans in algo_plan.required_translations.items():
+            bound.arguments[name] = self._add_translation_plan(
+                trans, bound.arguments[name]
+            )
+        args, kwargs = bound.args, bound.kwargs
+        # Add resolver if needed by the algorithm
+        if algo_plan.algo._include_resolver:
```
Comment (seibert) [inline metagraph/core/dask/resolver.py:117]: I noticed this code isn't covered by the tests, but when I tried writing a test, I discovered that the `include_resolver`  support for algorithms is not working in general.

**Review by seibert — CHANGES_REQUESTED** [review]
Overall this looks reasonable.  There are some followup enhancements we'll need to do.

**Comment by seibert** [comment]
I can think of some situations where will not know the types of the arguments we are passing to an algorithm, only that they are a Dask object.  For example, if an argument to a metagraph algorithm is coming from a `@delayed` function, then we have an interesting dispatching problem.  From that point forward, we would have to build a Dask graph of `abstract` algorithms that dispatch at execution time.  The translation steps would necessarily be not visible in the graph, and some kinds of exceptions that could have been generated earlier (like, there is no concrete implementation for the input types) would happen only at some point during execution.

**Comment by seibert** [comment]
#119 fixes the include_resolver bug.  Once that is merged, I can push some more tests for this PR.

File: metagraph/core/dask/resolver.py (hunk 1 above)
Comment (jim22k) [reply metagraph/core/dask/resolver.py:100]: Done.

**Comment by jim22k** [comment]
> I can think of some situations where will not know the types of the arguments we are passing to an algorithm, only that they are a Dask object. For example, if an argument to a metagraph algorithm is coming from a `@delayed` function, then we have an interesting dispatching problem. From that point forward, we would have to build a Dask graph of `abstract` algorithms that dispatch at execution time. The translation steps would necessarily be not visible in the graph, and some kinds of exceptions that could have been generated earlier (like, there is no concrete implementation for the input types) would happen only at some point during execution.

I created the DelayedWrapper class to handle this situation. If users wrap their constructor in `daskresolver.delayed_wrapper`, they can indicate the resultant type. Calling the delayed constructor will create a `Placeholder` object of the correct type.
